src/__tests__/lib/containers/entity/annotations/SchemaDrivenAnnotationsEditor.test.tsx

```diff
@@ -0,0 +1,459 @@
+      'country*',
+    )) as HTMLInputElement
+    const stateField = (await screen.findByLabelText(
+      'state*',
+    )) as HTMLInputElement
+    expect(countryField.value).toBe('USA')
+    expect(stateField.value).toBe('Washington')
+  })
+
+  it('Removes a custom annotation field when the last value is removed', async () => {
+    server.use(annotationsHandler, noSchemaHandler)
+    renderComponent()
+
+    // Should be able to find the 'country' custom field
+    await screen.findByRole('textbox', {
+      name: 'country-0',
+    })
+
+    // Remove the last element
+    userEvent.click(await screen.findByLabelText('Remove country-0'))
+
+    expect(
+      screen.queryByRole('textbox', {
+        name: 'country-0',
+      }),
+    ).not.toBeInTheDocument()
+  })
+
+  it('Sends a request to update annotations (no schema)', async () => {
+    server.use(annotationsHandler, noSchemaHandler, successfulUpdateHandler)
+
+    renderComponent()
+
+    await clickSave()
+
+    expect(updatedJsonCaptor).toBeCalledWith(
+      expect.objectContaining({ country: ['USA'], state: ['Washington'] }),
+    )
+  })
+
+  // TODO: This test is unstable, unclear why
+  it.skip('Sends a request to update annotations (with schema)', async () => {
+    // The annotations are predefined to match the schema
+    server.use(annotationsHandler, ...schemaHandlers, successfulUpdateHandler)
+    renderComponent()
+    await screen.findByText('requires scientific annotations', { exact: false })
+
+    // We need the individual schema field components to render because in some cases they convert data
+    // to the appropriate format (e.g. flattening arrays of one item)
+    await screen.findByLabelText('country*')
+    await screen.findByLabelText('state*')
+
+    await clickSave()
+
+    expect(updatedJsonCaptor).toBeCalledWith(
+      expect.objectContaining({ country: 'USA', state: 'Washington' }),
+    )
+  })
+
+  it('Prompts confirmation when the annotations are non-conformant', async () => {
+    // The schema requires annotations that are missing.
+    // Verify that we get a warning, and can still click through to save
+    server.use(noAnnotationsHandler, ...schemaHandlers, successfulUpdateHandler)
+    renderComponent()
+    await screen.findByText('requires scientific annotations', { exact: false })
+
+    await clickSaveAndConfirm()
+
+    await waitFor(() => expect(mockOnSuccessFn).toBeCalled())
+  })
+
+  it('Handles a failed update request', async () => {
+    server.use(annotationsHandler, noSchemaHandler, unsuccessfulUpdateHandler)
+
+    renderComponent()
+
+    const saveButton = await screen.findByRole('button', { name: 'Save' })
+    userEvent.click(saveButton)
+
+    await screen.findByText(
+      'Annotations could not be updated: Object was updated since last fetched',
+    )
+  })
+
+  it('Converts singular data to an additionalProperty array when removed from the schema', async () => {
```

Comment:
This and following tests cover cases where a conditional schema may cause data to change to/from `additionalProperties`. They test the resulting behavior of the data manipulation done in the custom field components passed to RJSF.